File: python/leet/codec.py
"""
Compact binary encoding for 1337.

Ultra-compact binary format with no repeated metadata.
Axes are identified by implicit order (0-31).

COGON format (96 fixed bytes):
    [HEADER][PAYLOAD][CHECKSUM]

    HEADER (4 bytes):
        - magic: 2 bytes (0x1337)
        - version_flags: 1 byte (version 4 bits + flags 4 bits)
        - reserved: 1 byte

    PAYLOAD (88 bytes):
        - id: 16 bytes (UUID)
        - sem: 32 bytes (32 uint8 values, fixed order A0-C10)
        - unc: 32 bytes (32 uint8 values, fixed order A0-C10)
        - stamp: 8 bytes (uint64, nanoseconds)

    CHECKSUM (4 bytes):
        - CRC32 of header + payload for integrity

Quantization:
    - Float [0.0, 1.0] → uint8: int(value * 255)
    - uint8 → Float: value / 255.0
    - Precision: ~0.4%

Compression vs JSON: ~5:1
Integrity: CRC32 verifies data corruption
"""

from __future__ import annotations

import struct
import zlib
from typing import Optional
from uuid import UUID

from leet.types import Cogon, Dag, Msg1337, Intent, Receiver, Surface, CanonicalSpace, Edge

# Constants
MAGIC = 0x1337
VERSION = 0x02  # Version 2 (with checksum)

# Header format: magic (2) + version_flags (1) + reserved (1)
HEADER_FORMAT = "!HBB"
HEADER_SIZE = 4

# Cogon format: id (16) + sem (32) + unc (32) + stamp (8)
COGON_FORMAT = "!16s32B32BQ"
COGON_SIZE = 88  # 16 + 32 + 32 + 8

# Checksum size
CHECKSUM_SIZE = 4

# Total binary size
TOTAL_BINARY_SIZE = HEADER_SIZE + COGON_SIZE + CHECKSUM_SIZE  # 96 bytes


def _compute_crc32(data: bytes) -> int:
    """Computes the CRC32 of the payload (unsigned)."""
    return zlib.crc32(data) & 0xFFFFFFFF


def _float_to_uint8(value: float) -> int:
    """Quantizes a float [0,1] to uint8 [0,255]."""
    return max(0, min(255, int(value * 255)))


def _uint8_to_float(value: int) -> float:
    """De-quantizes a uint8 [0,255] to float [0,1]."""
    return value / 255.0


def encode_cogon(cogon: Cogon) -> bytes:
    """
    Encodes a COGON to bytes (compact binary format with checksum).

    Returns:
        96 bytes (4 header + 88 payload + 4 checksum)
    """
    # Header
    version_flags = (VERSION << 4) | 0x00  # version 2, flags 0
    header = struct.pack(HEADER_FORMAT, MAGIC, version_flags, 0)

    # Payload
    uuid_bytes = UUID(cogon.id).bytes
    sem_bytes = [_float_to_uint8(v) for v in cogon.sem]
    unc_bytes = [_float_to_uint8(v) for v in cogon.unc]

    payload = struct.pack(COGON_FORMAT, uuid_bytes, *sem_bytes, *unc_bytes, cogon.stamp)

    # Compute CRC32 of header + payload
    crc = _compute_crc32(header + payload)
    checksum = struct.pack("!I", crc)

    return header + payload + checksum


def decode_cogon(data: bytes) -> Cogon:
    """
    Decodes bytes to a COGON with checksum validation.

    Args:
        data: 96 bytes in the binary format

    Raises:
        ValueError: If the format is invalid or the checksum fails
    """
    if len(data) < TOTAL_BINARY_SIZE:
        raise ValueError(
            f"Invalid data size: {len(data)} bytes, expected {TOTAL_BINARY_SIZE}"
        )

    # Split components
    header_data = data[:HEADER_SIZE]
    payload_data = data[HEADER_SIZE:HEADER_SIZE + COGON_SIZE]
    checksum_data = data[HEADER_SIZE + COGON_SIZE:HEADER_SIZE + COGON_SIZE + CHECKSUM_SIZE]

    # Parse header
    magic, version_flags, _ = struct.unpack(HEADER_FORMAT, header_data)

    if magic != MAGIC:
        raise ValueError(f"Invalid magic: 0x{magic:04x}, expected 0x{MAGIC:04x}")

    version = version_flags >> 4
    if version != VERSION:
        raise ValueError(f"Unsupported version: {version}, expected {VERSION}")

    # Verify checksum
    stored_crc = struct.unpack("!I", checksum_data)[0]
    computed_crc = _compute_crc32(header_data + payload_data)

    if stored_crc != computed_crc:
        raise ValueError(
            f"Checksum mismatch: stored=0x{stored_crc:08x}, computed=0x{computed_crc:08x}. "
            "Data may be corrupted."
        )

    # Parse payload
    unpacked = struct.unpack(COGON_FORMAT, payload_data)

    uuid_bytes = unpacked[0]
    sem_bytes = unpacked[1:33]  # 32 bytes
    unc_bytes = unpacked[33:65]  # 32 bytes
    stamp = unpacked[65]

    # Convert back
    cogon_id = str(UUID(bytes=uuid_bytes))
    sem = [_uint8_to_float(v) for v in sem_bytes]
    unc = [_uint8_to_float(v) for v in unc_bytes]

    return Cogon(
        id=cogon_id,
        sem=sem,
        unc=unc,
        stamp=stamp,
        raw=None  # RAW is not supported in the compact binary format
    )


def encode_msg(msg: Msg1337) -> bytes:
    """
    Encodes a MSG_1337 to bytes.

    Format:
        [HEADER][COGON PAYLOAD][MSG EXTENSION]

    MSG EXTENSION:
        - sender_len: 1 byte
        - sender: N bytes
        - receiver_len: 1 byte (0 = broadcast)
        - receiver: N bytes (omitted if broadcast)
        - intent: 1 byte
        - has_patch: 1 byte (0 or 1)
        - patch: 32 bytes * 4 (if has_patch=1, 32 floats)
    """
    # Encode base Cogon payload
    base = encode_cogon(msg.payload if hasattr(msg.payload, 'sem') else msg.payload.nodes[0])

    # MSG extension
    sender_bytes = msg.sender.encode('utf-8')
    sender_len = len(sender_bytes)

    receiver_id = msg.receiver.agent_id if msg.receiver.agent_id else ""
    receiver_bytes = receiver_id.encode('utf-8')
    receiver_len = len(receiver_bytes)

    intent_value = Intent(msg.intent).value if isinstance(msg.intent, str) else msg.intent.value
    intent_code = {
        "ASSERT": 0, "QUERY": 1, "DELTA": 2, "SYNC": 3, "ANOMALY": 4, "ACK": 5
    }.get(intent_value, 0)

    has_patch = 1 if msg.patch else 0

    # Build extension
    ext = struct.pack("!B", sender_len)
    ext += sender_bytes
    ext += struct.pack("!B", receiver_len)
    ext += receiver_bytes
    ext += struct.pack("!BB", intent_code, has_patch)

    if msg.patch:
        # Pack 32 floats as 32 uint8 (quantized)
        patch_bytes = [_float_to_uint8(v) for v in msg.patch]
        ext += struct.pack("!32B", *patch_bytes)

    return base + ext


def decode_msg(data: bytes) -> Msg1337:
    """Decodes bytes to a MSG_1337."""
    # First decode the Cogon part
    cogon = decode_cogon(data)

    # Parse extension
    offset = HEADER_SIZE + COGON_SIZE

    sender_len = data[offset]
    offset += 1
    sender = data[offset:offset + sender_len].decode('utf-8')
    offset += sender_len

    receiver_len = data[offset]
    offset += 1
    receiver_id = data[offset:offset + receiver_len].decode('utf-8') if receiver_len > 0 else None
    offset += receiver_len

    intent_code, has_patch = struct.unpack("!BB", data[offset:offset + 2])
    offset += 2

    intent_map = {0: "ASSERT", 1: "QUERY", 2: "DELTA", 3: "SYNC", 4: "ANOMALY", 5: "ACK"}
    intent = intent_map.get(intent_code, "ASSERT")

    patch = None
    if has_patch:
        patch_bytes = struct.unpack("!32B", data[offset:offset + 32])
        patch = [_uint8_to_float(v) for v in patch_bytes]

    return Msg1337(
        id=cogon.id,  # Reuse cogon ID for message
        sender=sender,
        receiver=Receiver(agent_id=receiver_id),
        intent=intent,
        payload=cogon,
        c5=CanonicalSpace(
            zone_fixed=[0.5] * 32,
            zone_emergent={},
            schema_ver="0.4",
            align_hash=""
        ),
        surface=Surface(
            human_required=False,
            urgency=None,
            reconstruct_depth=0,
            lang="pt"
        ),
        patch=patch
    )


def get_binary_size(cogon: Cogon) -> int:
    """Returns the size in bytes of the binary encoding."""
    return TOTAL_BINARY_SIZE


def compare_sizes(cogon: Cogon) -> dict:
    """Compares sizes between JSON and binary."""
    json_size = len(cogon.to_json().encode('utf-8'))
    binary_size = get_binary_size(cogon)

    return {
        "json_bytes": json_size,
        "binary_bytes": binary_size,
        "compression_ratio": json_size / binary_size,
        "space_saved_percent": (1 - binary_size / json_size) * 100
    }


# Add methods to Cogon for convenience
def _cogon_to_bytes(self) -> bytes:
    """Export COGON to binary format."""
    return encode_cogon(self)


def _cogon_from_bytes(cls, data: bytes) -> Cogon:
    """Import COGON from binary format."""
    return decode_cogon(data)


# Monkey-patch Cogon
Cogon.to_bytes = _cogon_to_bytes
Cogon.from_bytes = classmethod(_cogon_from_bytes)


__all__ = [
    'encode_cogon',
    'decode_cogon',
    'encode_msg',
    'decode_msg',
    'get_binary_size',
    'compare_sizes',
]

File: leet-core/src/codec.rs
//! Binary codec for 1337 — compact binary encoding with checksum.
//!
//! Format COGON (96 bytes fixed):
//! ```text
//! [HEADER: 4 bytes][PAYLOAD: 88 bytes][CHECKSUM: 4 bytes]
//!
//! Header:
//!   - magic: 2 bytes (0x1337)
//!   - version_flags: 1 byte
//!   - reserved: 1 byte
//!
//! Payload:
//!   - id: 16 bytes (UUID)
//!   - sem: 32 bytes (32 uint8, quantized)
//!   - reserved: 32 bytes (zeros; was unc in v0.4)  // TODO(07c): rename to reserved
//!   - stamp: 8 bytes (u64 nanoseconds)
//!
//! Checksum:
//!   - CRC32 of header + payload for integrity verification
//! ```
//!
//! Quantization: float [0.0, 1.0] <-> uint8 [0, 255]
//! Compression vs JSON: ~4-5:1
//! Integrity: CRC32 detects data corruption

use crate::types::{Cogon, SemVec};
use uuid::Uuid;

/// Magic number for binary format
pub const MAGIC: u16 = 0x1337;
/// Version 2 (with checksum)
pub const VERSION: u8 = 0x02;
/// Header size in bytes
pub const HEADER_SIZE: usize = 4;
/// Payload size in bytes
pub const PAYLOAD_SIZE: usize = 88;
/// Checksum size in bytes
pub const CHECKSUM_SIZE: usize = 4;
/// Total binary size
pub const TOTAL_SIZE: usize = HEADER_SIZE + PAYLOAD_SIZE + CHECKSUM_SIZE; // 96 bytes

/// Compute CRC32 of data
#[inline]
pub fn compute_crc32(data: &[u8]) -> u32 {
    crc32fast::hash(data)
}

/// Quantize float [0.0, 1.0] to uint8 [0, 255]
#[inline]
pub fn float_to_u8(v: f32) -> u8 {
    (v.clamp(0.0, 1.0) * 255.0).round() as u8
}

/// De-quantize uint8 [0, 255] to float [0.0, 1.0]
#[inline]
pub fn u8_to_float(v: u8) -> f32 {
    v as f32 / 255.0
}

/// Encode a Cogon to binary format (96 bytes with checksum)
pub fn encode_cogon(cogon: &Cogon) -> Vec<u8> {
    let mut buf = Vec::with_capacity(TOTAL_SIZE);

    // Header
    buf.extend_from_slice(&MAGIC.to_be_bytes());
    buf.push((VERSION << 4) | 0x00); // version + flags
    buf.push(0x00); // reserved

    // Payload: UUID (16 bytes)
    buf.extend_from_slice(cogon.id.as_bytes());

    // Payload: sem (32 bytes)
    for &v in &cogon.sem {
        buf.push(float_to_u8(v));
    }

    // Payload: reserved 32 bytes (zeros; was unc in v0.4 — kept for wire compat)
    buf.extend_from_slice(&[0u8; 32]);

    // Payload: stamp (8 bytes, big-endian u64)
    buf.extend_from_slice(&cogon.stamp.to_be_bytes());

    // Checksum: CRC32 of header + payload
    let crc = compute_crc32(&buf);
    buf.extend_from_slice(&crc.to_be_bytes());

    buf
}

/// Decode a Cogon from binary format with checksum verification
pub fn decode_cogon(data: &[u8]) -> Result<Cogon, crate::error::LeetError> {
    if data.len() < TOTAL_SIZE {
        return Err(crate::error::LeetError::SerializationError(
            format!("Invalid data size: {} bytes, expected {}", data.len(), TOTAL_SIZE)
        ));
    }

    // Split components
    let header_data = &data[..HEADER_SIZE];
    let payload_data = &data[HEADER_SIZE..HEADER_SIZE + PAYLOAD_SIZE];
    let checksum_data = &data[HEADER_SIZE + PAYLOAD_SIZE..HEADER_SIZE + PAYLOAD_SIZE + CHECKSUM_SIZE];

    // Parse header
    let magic = u16::from_be_bytes([header_data[0], header_data[1]]);
    if magic != MAGIC {
        return Err(crate::error::LeetError::SerializationError(
            format!("Invalid magic: 0x{:04x}, expected 0x{:04x}", magic, MAGIC)
        ));
    }

    let version = header_data[2] >> 4;
    if version != VERSION {
        return Err(crate::error::LeetError::SerializationError(
            format!("Unsupported version: {}, expected {}", version, VERSION)
        ));
    }

    // Verify checksum
    let stored_crc = u32::from_be_bytes([
        checksum_data[0], checksum_data[1],
        checksum_data[2], checksum_data[3]
    ]);
    let computed_crc = compute_crc32(&data[..HEADER_SIZE + PAYLOAD_SIZE]);

    if stored_crc != computed_crc {
        return Err(crate::error::LeetError::SerializationError(
            format!(
                "Checksum mismatch: stored=0x{:08x}, computed=0x{:08x}. Data may be corrupted.",
                stored_crc, computed_crc
            )
        ));
    }

    // Parse payload
    let id = Uuid::from_bytes([
        payload_data[0], payload_data[1], payload_data[2], payload_data[3],
        payload_data[4], payload_data[5], payload_data[6], payload_data[7],
        payload_data[8], payload_data[9], payload_data[10], payload_data[11],
        payload_data[12], payload_data[13], payload_data[14], payload_data[15],
    ]);

    // sem (32 bytes at offset 16)
    let mut sem: SemVec = [0.0; 32];
    for i in 0..32 {
        sem[i] = u8_to_float(payload_data[16 + i]);
    }

    // reserved 32 bytes (offset 48–79) — ignored

    // stamp (8 bytes at offset 80)
    let stamp = i64::from_be_bytes([
        payload_data[80], payload_data[81], payload_data[82], payload_data[83],
        payload_data[84], payload_data[85], payload_data[86], payload_data[87],
    ]);

    Ok(Cogon {
        id,
        sem,
        stamp,
        raw: None,
    })
}

/// Get the size of binary encoding for a Cogon
pub fn binary_size(_cogon: &Cogon) -> usize {
    TOTAL_SIZE
}

/// Compare sizes between JSON (approximate) and binary
pub fn compare_sizes(cogon: &Cogon) -> SizeComparison {
    let binary_bytes = binary_size(cogon);
    let json_bytes = estimate_json_size(cogon);

    SizeComparison {
        json_bytes,
        binary_bytes,
        compression_ratio: json_bytes as f32 / binary_bytes as f32,
        space_saved_percent: (1.0 - binary_bytes as f32 / json_bytes as f32) * 100.0,
    }
}

/// Estimate JSON serialization size
fn estimate_json_size(_cogon: &Cogon) -> usize {
    // Rough estimate: UUID (~36) + sem array (~200) + metadata
    36 + 32 * 6 + 100
}

/// Size comparison result
#[derive(Debug, Clone, Copy)]
pub struct SizeComparison {
    pub json_bytes: usize,
    pub binary_bytes: usize,
    pub compression_ratio: f32,
    pub space_saved_percent: f32,
}

impl Cogon {
    /// Export COGON to binary format
    pub fn to_bytes(&self) -> Vec<u8> {
        encode_cogon(self)
    }

    /// Import COGON from binary format
    pub fn from_bytes(data: &[u8]) -> Result<Self, crate::error::LeetError> {
        decode_cogon(data)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_cogon(sem_val: f32) -> Cogon {
        Cogon {
            id: Uuid::new_v4(),
            sem: [sem_val; 32],
            stamp: 0,
            raw: None,
        }
    }

    #[test]
    fn test_quantization_roundtrip() {
        let original = 0.734;
        let quantized = float_to_u8(original);
        let recovered = u8_to_float(quantized);
        assert!((original - recovered).abs() < 0.004);
    }

    #[test]
    fn test_encode_decode_roundtrip() {
        let original = make_cogon(0.5);
        let encoded = encode_cogon(&original);
        let decoded = decode_cogon(&encoded).unwrap();

        assert_eq!(original.id, decoded.id);
        for i in 0..32 {
            assert!((original.sem[i] - decoded.sem[i]).abs() < 0.004);
        }
    }

    #[test]
    fn test_cogon_zero_roundtrip() {
        let zero = Cogon::zero();
        let encoded = encode_cogon(&zero);
        let decoded = decode_cogon(&encoded).unwrap();
        assert_eq!(zero.id, decoded.id);
        // Quantization tolerance: 1/255 ≈ 0.004
        for i in 0..32 {
            assert!((zero.sem[i] - decoded.sem[i]).abs() < 0.004,
                "sem[{}]: {} vs {}", i, zero.sem[i], decoded.sem[i]);
        }
    }

    #[test]
    fn test_binary_size_constant() {
        let c1 = make_cogon(0.1);
        let c2 = make_cogon(0.9);
        assert_eq!(binary_size(&c1), TOTAL_SIZE);
        assert_eq!(binary_size(&c2), TOTAL_SIZE);
    }

    #[test]
    fn test_invalid_magic() {
        let mut data = encode_cogon(&make_cogon(0.5));
        data[0] = 0x00;
        data[1] = 0x00;
        assert!(decode_cogon(&data).is_err());
    }

    #[test]
    fn test_invalid_size() {
        let data = vec![0x13, 0x37, 0x20, 0x00]; // Too small
        assert!(decode_cogon(&data).is_err());
    }

    #[test]
    fn test_checksum_mismatch() {
        let mut data = encode_cogon(&make_cogon(0.5));
        data[10] ^= 0xFF;
        let result = decode_cogon(&data);
        assert!(result.is_err());
        let err_msg = format!("{}", result.unwrap_err());
        assert!(err_msg.contains("Checksum mismatch"));
    }

    #[test]
    fn test_compression_ratio() {
        let cogon = make_cogon(0.5);
        let comparison = compare_sizes(&cogon);
        assert!(comparison.compression_ratio > 3.0);
        assert!(comparison.space_saved_percent > 60.0);
    }

    #[test]
    fn test_cogon_to_bytes_method() {
        let cogon = make_cogon(0.5);
        let data = cogon.to_bytes();
        assert_eq!(data.len(), TOTAL_SIZE);
    }

    #[test]
    fn test_cogon_from_bytes_method() {
        let original = make_cogon(0.5);
        let data = original.to_bytes();
        let recovered = Cogon::from_bytes(&data).unwrap();
        assert_eq!(original.id, recovered.id);
    }
}
